src/main/java/com/startsteps/ecommerceapi/user/service/UserServiceImpl.java

```diff
@@ -1,42 +1,134 @@
+import java.time.temporal.Temporal;
+import java.util.Date;
+import java.util.NoSuchElementException;
 import java.util.Optional;
+import java.util.UUID;
 
 /**
  * service layer class for serving "api/users" endpoints
  */
+
 @Service
 @Slf4j
+@Transactional
 public class UserServiceImpl implements UserService {
     private final UserRepository userRepository;
-
-    public UserServiceImpl(UserRepository userRepository) {
+    private final PasswordResetTokenRepository passwordResetTokenRepository;
+    private final PasswordEncoder passwordEncoder;
+    private final EmailService emailService;
+   @Autowired
+   public UserServiceImpl(UserRepository userRepository, PasswordResetTokenRepository passwordResetTokenRepository, PasswordEncoder passwordEncoder, EmailService emailService) {
         this.userRepository = userRepository;
-    }
+       this.passwordResetTokenRepository = passwordResetTokenRepository;
+       this.passwordEncoder = passwordEncoder;
+       this.emailService = emailService;
+   }
 
+    @Override
+    public Optional<User> findUserByEmail(String emailAddress) {
+        return userRepository.findByEmail(emailAddress);
+    }
 
     @Override
+    public Optional<User> findUserByUsername(String username) {
+        return userRepository.findByUsername(username);
+    }
 
+    @Override
+    public User registerUser(UserDTO user) {
+        if(userRepository.existsByEmail(user.getEmail())){
+            throw new UserAlreadyExistsException("This email " + user.getEmail() + " already exists");
+        }
+        else if(userRepository.existsByUsername(user.getUsername())){
+            throw new UserAlreadyExistsException("This username " + user.getUsername() + " is already used");
+        }
+        User newUser = new User((user.getUserId()), user.getUsername(), passwordEncoder.encode(user.getPassword()),
+                user.getEmail(),false, true);
+        return userRepository.save(newUser);
+    }
+    @Override //TODO: authorized method
+    public User registerAdmin(UserDTO user) {
+        return null;
+    }
 
     @Override
-    public List<User> findAllUsers() {
-        return userRepository.findAll();
+    public User resetPasswordSendEmail(String userEmail) {
+        if(!userRepository.existsByEmail(userEmail) ) {
+            throw new UserNotFoundException("User with the email " + userEmail + " not found");
+        }
+        User user = userRepository.findByEmail(userEmail).orElseThrow(() ->
+                new UserNotFoundException("User with the email " + userEmail + " not found"));
+        String resetToken = generateResetToken();
+        createPasswordResetTokenForUser(user, resetToken);
+        SimpleMailMessage email = new SimpleMailMessage();
+        email.setTo(userEmail);
+        email.setSubject("Reset your password");
+        email.setText("To reset your password, click the link below:\n" +
+                "http://ecommerce-api/reset-password?token=" + resetToken);
```

Comment:
this won't work until you release and have a proper ip address/domain name but it's cool to see it